Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,381 @@
+      rw [Nat.add_sub_cancel' hi.le]
+  -- We need to show that for all `α ∈ s`, the polynomial `coeff χ i` evaluates to zero at `α`.
+  intro α hα
+  -- Once again, we are left with showing that `⁅y, _⁆` acts nilpotently on `E`.
+  rw [← coe_evalRingHom, ← coeff_map, lieCharpoly_map_eval,
+    (LinearMap.charpoly_eq_X_pow_iff _).mpr, coeff_X_pow, if_neg hi.ne]
+  -- To do so, it suffices to show that the Engel subalgebra of `y = a • u + x` is contained in `E`.
+  let y := α • u + x
+  suffices engel K (y : L) ≤ engel K (x : L) by
+    -- Indeed, in that case the minimality assumption on `E` implies
+    -- that `E` is contained in the Engel subalgebra of `y`.
+    replace this : engel K (x : L) ≤ engel K (y : L) := (hmin y this).ge
+    intro z
+    -- And so we are done, by the definition of Engel subalgebra.
+    simpa only [mem_engel_iff, Subtype.ext_iff, LieSubmodule.coe_toEndomorphism_pow] using this z.2
+  -- Now we are in good shape.
+  -- Fix an element `z` in the Engel subalgebra of `y`.
+  intro z hz
+  -- We need to show that `z` is in `E`, or alternatively that `z = 0` in `Q`.
+  show z ∈ E
+  rw [← LieSubmodule.Quotient.mk_eq_zero]
+  -- We denote the image of `z` in `Q` by `z'`.
+  set z' : Q := LieSubmodule.Quotient.mk' E z
+  -- First we observe that `z'` is killed by a power of `⁅y, _⁆`.
+  have hz' : ∃ n : ℕ, ((toEndomorphism K U Q) y ^ n) z' = 0 := by
+    rw [mem_engel_iff] at hz
+    obtain ⟨n, hn⟩ := hz
+    use n
+    apply_fun LieSubmodule.Quotient.mk' E at hn
+    rw [LieModuleHom.map_zero] at hn
+    rw [← hn]
+    clear hn
+    induction n with
+    | zero => simp only [Nat.zero_eq, pow_zero, LinearMap.one_apply]
+    | succ n ih => rw [pow_succ', pow_succ', LinearMap.mul_apply, ih]; rfl
+  classical
+  -- Now let `n` be the smallest power such that `⁅y, _⁆ ^ n` kills `z'`.
+  set n := Nat.find hz' with _hn
+  have hn : ((toEndomorphism K (↥U) Q) y ^ n) z' = 0 := Nat.find_spec hz'
+  -- If `n = 0`, then we are done.
+  obtain hn₀|⟨k, hk⟩ : n = 0 ∨ ∃ k, n = k + 1 := by cases n <;> simp
+  · simpa only [hn₀, pow_zero, LinearMap.one_apply] using hn
+  -- If `n = k + 1`, then we can write `⁅y, _⁆ ^ n = ⁅y, _⁆ ∘ ⁅y, _⁆ ^ k`.
+  -- Recall that `constantCoeff ψ` is non-zero on `α`, and `y = α • u + x`.
+  specialize hsψ α hα
+  -- Hence `⁅y, _⁆` acts injectively on `Q`.
+  rw [← coe_evalRingHom, constantCoeff_apply, ← coeff_map, lieCharpoly_map_eval,
+    ← constantCoeff_apply, ne_eq, LinearMap.charpoly_constantCoeff_eq_zero_iff] at hsψ
+  -- We deduce from this that `z' = 0`, arguing by contraposition.
+  contrapose! hsψ
+  -- Indeed `⁅y, _⁆` kills `⁅y, _⁆ ^ k` applied to `z'`.
+  use ((LieModule.toEndomorphism K U Q) y ^ k) z'
+  refine ⟨?_, ?_⟩
+  · -- And `⁅y, _⁆ ^ k` applied to `z'` is non-zero by definition of `n`.
+    apply Nat.find_min hz'; omega
+  · rw [← hn, hk, pow_succ', LinearMap.mul_apply]
+
+variable (K L)
+
+open Cardinal in
+lemma exists_IsCartanSubalgebra_of_finrank_le_card (h : finrank K L ≤ #K) :
```

Comment:
I think `exists_isCartanSubalgebra_of_finrank_le_card` with lower-case `i`, also in the next lemma

Reply:
Thanks, fixed